invenio_jobs/logging/jobs.py

```diff
@@ -0,0 +1,136 @@
+    """Custom logging handler that enriches logs with global context and indexes them in OS."""
+
+    def emit(self, record):
+        """Emit log record after enriching it with global context."""
+        if job_context.get():
+            enriched_log = self.enrich_log(record)
+            self.index_in_os(enriched_log)
+
+    def enrich_log(self, record):
+        """Enrich log record with contextvars' global context."""
+        log_data = {
+            "timestamp": datetime.now().isoformat(),
+            "level": record.levelname,
+            "message": record.getMessage(),
+            "module": record.module,
+            "function": record.funcName,
+            "line": record.lineno,
+            "context": job_context.get(),
+        }
+        serialzed_data = JobLogEntrySchema().load(log_data)
+        return serialzed_data
+
+    def index_in_os(self, log_data):
+        """Send log data to OpenSearch."""
+        index_prefix = current_app.config.get("SEARCH_INDEX_PREFIX", "")
+        full_index_name = f"{index_prefix}job-logs"
+        current_search_client.index(index=full_index_name, body=log_data)
+
+
+@contextmanager
+def set_job_context(data):
+    """Context manager for safely setting and cleaning up contextvars."""
+    token = job_context.set(data)
+    try:
+        yield job_context  # Yield the contextvar for modification
+    finally:
+        job_context.reset(token)  # Ensures cleanup
+
+
+def with_job_context(base_context={}):
+    """Decorator to apply and update context dynamically inside a function."""
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            with set_job_context(base_context) as context:
+
+                def update_context(new_data):
+                    """Update context dynamically within the function."""
+                    context.set({**context.get(), **new_data})
+
+                return func(
+                    *args, update_context=update_context, **kwargs
+                )  # Inject update_context
+
+        return wrapper
+
+    return decorator
+
+
+class InvenioLoggingJobs(InvenioLoggingBase):
+    """Logging extension for jobs."""
+
+    def init_app(self, app):
+        """Flask application initialization."""
+        self.init_config(app)
+
+        if not app.config["LOGGING_JOBS"]:
+            return
+        # For the jobs handler to log at the configured level
+        # we need to check the configured level and set it if it is lower than the app logger level
+        # otherwise the handler will not log anything
+        configured_level_str = app.config["LOGGING_JOBS_LEVEL"]
+        if configured_level_str is not None:
+            configured_level = logging._nameToLevel.get(configured_level_str.upper())
+            if (
+                isinstance(configured_level, int)
+                and app.logger.level > configured_level
+            ):
+                app.logger.setLevel(configured_level)
```

Comment:
Alternative solution to discuss:
We can have a default logger.leve set to DEBUG (minimun level) and let each handler define the level to log content

Reply:
Agreed to use DEBUG level for logger as default and configure the handlers level